engine/lib/luajit-ffi-gen/readme.md

```diff
@@ -111,9 +111,77 @@ for test:
 cargo expand -p luajit-ffi-gen --test basic_test
 ```
 
+## Supported types
+
+**luajit_ffi_gen** attribute supports different types as input parameters and return type in the impl methods.
+
+Following table shows representation of Rust types in the generated code.
+
+| Rust type                            | extern "C" interface         | C type |
+| ------------------------------------ | ------------------- | ---------- |
+| Immutable reference (&T)             | &T                  | T const*   |
+| Mutable reference (&mut T)           | &mut T              | T *        |
+| Self (in return position)            | Box\<T>             | T *        |
+| Basic and copy types (T)             | T                   | T          |
+| String, str                          | *const libc::c_char | cstr       |
+| Option\<T>                           | *mut T              | T *        |
+| Result\<T, E> (only return position) | T, panic on error   | T          |
```

Comment:
I wonder whether it would be nice to generate a C struct like:

`Result<T, E>` becomes
```
struct Result_T_E {
  T ok;
  E error;
  bool is_ok;
};
```

then in the Lua FFI code, we transform that struct into either `ok, nil`, or return two values `nil, error`, something kind of like what Roblox does: https://roblox.github.io/lua-style-guide/#error-handling

Reply:
Good idea :+1: 
I was thinking about using lua_error() function to return an error back to Lua but unfortunately it will require passing lua_State everywhere.
Your approach is better.
This will require some work so we should do it in another PR.

Reply:
Agreed